### PR TITLE
metadata initiation using REDCap is updated to be more effective

### PR DESCRIPTION
**Pull request by kcho**
Now REDCap API has been updated to specify the correct arms and fields to only pull the relevant information.

### CODE REVIEW

File: lochness/transfer/__init__.py [hunk 1]
```diff
@@ -375,6 +375,8 @@ def create_s3_transfer_table(Lochness, rewrite=False) -> None:
                     continue
                 try:
                     source = re.search(r'upload: (\S+)', line).group(1)
+                    # if not source.startswith('/'):
+                        # source = '/' + source
```
Comment (tashrifbillah) [inline lochness/transfer/__init__.py:379]: I would remove this (unintentional) dead code.

File: lochness/email/__init__.py [hunk 2]
```diff
@@ -75,7 +75,12 @@ def send_detail(Lochness,
 
     server_name = Lochness['project_name'] \
         if 'project_name' in Lochness else 'Data aggregation server'
-    title = f'{server_name}: {title} {datetime.now(tz).date()}'
+
+    if Lochness.get('production', False):
+        title = f'{server_name} Production: {title} {datetime.now(tz).date()}'
+    else:
+        title = f'{server_name}: {title} {datetime.now(tz).date()}'
```
Comment (tashrifbillah) [inline lochness/email/__init__.py:82]: Why don't we add a `Development` string here?

File: lochness/redcap/__init__.py [hunk 3]
```diff
@@ -72,117 +80,133 @@ def initialize_metadata(Lochness: 'Lochness object',
...
-            continue
+    # if empty REDCap
+    if len(df) == 0:
+        logger.warn(f'There are no records for {site_code_study}')
+        remove_file_that_may_exist(metadata_study)
+        return
 
-        subject_dict = {'Subject ID': item[redcap_id_colname]}
+    # only keep AMPSCZ rows
+    df = df[df[redcap_id_colname].str.match('[A-Z]{2}\d{5}')]
```
Comment (tashrifbillah) [inline lochness/redcap/__init__.py:136]: Does `\d{5}` mean five numbers?

File: lochness/redcap/__init__.py [hunk 4]
```diff
@@ -72,117 +80,133 @@ def initialize_metadata(Lochness: 'Lochness object',
...
+
+    # if no data matches AMPSCZ ID
+    if len(df) == 0:
+        logger.warn(f'There are no records for {site_code_study}')
+        remove_file_that_may_exist(metadata_study)
+        return
+
+    # make a single row for each subject record
+    # redcap_event_name column contains timepoint information (different arms)
+    df.drop('redcap_event_name', axis=1, inplace=True)
```
Comment (tashrifbillah) [inline lochness/redcap/__init__.py:146]: `redcap_event_name` must exist in the downloaded JSON:

image

If you dropping will remove it from the JSON, please undo this.

**Review by tashrifbillah — COMMENTED** [review]
Tashrif has asked a few clarifying questions. Thank you for all the complicated Lochness work.

File: lochness/utils/source_check.py [hunk 5]
```diff
@@ -362,45 +425,76 @@ def check_source(Lochness: 'lochness', test: bool = False) -> None:
...
+                          'Format']
+    
+    # to highlight REDCap table separately in the email, since this is the
+    # major bottleneck for the smooth dataflow
+    tmp = qc_fail_df[
+        (qc_fail_df['Consent date in DB'] == 'Consent date missing') |
+        (qc_fail_df['Subject ID in database'] == 'Correct')
+            ]
+    tmp['File Path'] = 'REDCap'
+    tmp['Data Type'] = 'REDCap'
```
[the remark this answers is not in the document]
Comment (kcho) [reply lochness/utils/source_check.py:488]: This is to include `REDCap` to be shown in the `File Path` column (REDCap source do not have file path)

File: lochness/transfer/__init__.py (hunk 1 above)
Comment (kcho) [reply lochness/transfer/__init__.py]: done

**Comment by tashrifbillah** [comment]
Unaddressed comment:
https://github.com/AMP-SCZ/lochness/pull/105#issuecomment-1165638798

File: lochness/redcap/__init__.py [hunk 6]
```diff
@@ -72,117 +80,133 @@ def initialize_metadata(Lochness: 'Lochness object',
...
-                    }
+    # to extract ID and consent form for all the records that
+    # belong to the site from screening & baseline arms
+    record_query = {
+        'token': api_key,
+        'content': 'record',
+        'format': 'json',
+        'fields[0]': redcap_id_colname,
+        'fields[1]': redcap_consent_colname,
+        'events[0]': 'screening_arm_1',
```
Comment (kcho) [inline lochness/redcap/__init__.py:102]: @tashrifbillah look here

Comment (kcho) [reply lochness/redcap/__init__.py:102]: https://github.com/AMP-SCZ/lochness/pull/105#issuecomment-1165680463

Comment (tashrifbillah) [reply lochness/redcap/__init__.py:102]: Okay I see it now--you have hard-coded it as opposed to your description hinting it has been made user-specifiable:

> has been updated to specify the correct arms

**Review by tashrifbillah — APPROVED** [review]

**Pull request merged**